# Data Science Project 

This project will analyse a data set about used cars from ebay in germany.

In [3]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")


In [4]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [6]:
autos.info()
print("\n")
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Some observations about data:
- Columns names are in camel case
- Some of the fields carry null values
- Some columns need cleaning and they could be converted to Int/Float/Date

In [7]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [8]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_Type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']

autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_Type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Have changed the column names from Camel case to remain consistent convention which makes it easier to read and understand code.

In [10]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-05 16:57:05  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_Type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [20]:
print(autos["nr_of_pictures"].value_counts())
print(autos["postal_code"].value_counts())

0    50000
Name: nr_of_pictures, dtype: int64
10115    109
65428    104
66333     54
45888     50
44145     48
        ... 
23942      1
83365      1
95683      1
97794      1
67585      1
Name: postal_code, Length: 7014, dtype: int64


- Mostly single frequency (it's better to be dropped) -: nr_of_pictures
- Columns that needs further investigations: name
- Numeric Columns present as string: price, odometer

In [31]:
cleaned_price = autos["price"].str.replace("$","").str.replace(",","").astype(int)
cleaned_odometer = autos["odometer"].str.replace("km","").str.replace(",","").astype(int)
autos["price"] = cleaned_price
autos["odometer"] = cleaned_odometer
print(autos["price"].dtype)

autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)
print(autos["odometer_km"].dtype)


int64
int64


In [32]:
min_price = autos["price"].min()
max_price = autos["price"].max()
print("Min Price: " + str(min_price) + "\n" + "Max Price: " + str(max_price))
min_dist = autos["odometer_km"].min()
max_dist = autos["odometer_km"].max()
print("Min Dist: " + str(min_dist) + "\n" + "Max Dist: " + str(max_dist))

Min Price: 0
Max Price: 99999999
Min Dist: 5000
Max Dist: 150000


In [36]:
print(autos["price"].unique().shape[0])
print(autos["odometer_km"].unique().shape[0])
autos["price"].describe()

2357
13


count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [45]:
#print(autos["price"].value_counts().sort_index(ascending=True).head(100))
autos = autos[autos["price"].between(1000,13000000)]
autos.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 38638 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        38638 non-null  object
 1   name                38638 non-null  object
 2   seller              38638 non-null  object
 3   offer_type          38638 non-null  object
 4   price               38638 non-null  int64 
 5   abtest              38638 non-null  object
 6   vehicle_type        35879 non-null  object
 7   registration_year   38638 non-null  int64 
 8   gearbox             37200 non-null  object
 9   power_ps            38638 non-null  int64 
 10  model               37015 non-null  object
 11  odometer_km         38638 non-null  int64 
 12  registration_month  38638 non-null  int64 
 13  fuel_Type           36275 non-null  object
 14  brand               38638 non-null  object
 15  unrepaired_damage   32863 non-null  object
 16  ad_created          38

The original data set contained some unrealistic prices ranging from $0 to $99 million. I have removed all the aoutliers and shorthened the data to approximately 39000 rows.

In [57]:
date_crawled = autos["date_crawled"].str[:10].value_counts(normalize=True, ascending=True, dropna=False)*100
date_crawled

2016-04-07    0.150111
2016-04-06    0.326104
2016-03-18    1.283710
2016-04-05    1.335473
2016-03-06    1.387235
2016-03-13    1.599462
2016-03-05    2.554480
2016-03-24    2.901289
2016-03-16    2.906465
2016-03-17    3.048812
2016-03-25    3.051400
2016-03-27    3.139396
2016-03-31    3.141985
2016-03-23    3.219628
2016-03-09    3.248098
2016-03-22    3.253274
2016-03-08    3.263626
2016-03-11    3.279155
2016-03-30    3.305036
2016-03-26    3.310213
2016-03-10    3.330918
2016-03-15    3.361975
2016-03-29    3.400797
2016-04-01    3.460324
2016-03-07    3.512087
2016-03-19    3.512087
2016-03-28    3.535380
2016-04-02    3.628552
2016-03-14    3.662198
2016-04-04    3.690667
2016-03-21    3.732077
2016-03-12    3.739842
2016-03-20    3.814897
2016-04-03    3.913246
Name: date_crawled, dtype: float64

In [59]:
ad_created = autos["ad_created"].str[:10].value_counts(normalize=True, ascending=True, dropna=False)*100
ad_created

2016-02-11    0.002588
2016-02-07    0.002588
2015-06-11    0.002588
2016-01-03    0.002588
2016-01-29    0.002588
                ...   
2016-03-12    3.713960
2016-04-04    3.732077
2016-03-21    3.760547
2016-03-20    3.825250
2016-04-03    3.944304
Name: ad_created, Length: 74, dtype: float64

In [60]:
last_seen = autos["last_seen"].str[:10].value_counts(normalize=True, ascending=True, dropna=False)*100
last_seen

2016-03-05     0.108701
2016-03-06     0.357161
2016-03-07     0.455510
2016-03-08     0.623738
2016-03-18     0.737616
2016-03-13     0.838553
2016-03-09     0.892903
2016-03-10     0.980900
2016-03-11     1.172421
2016-03-14     1.198302
2016-03-27     1.407940
2016-03-19     1.459703
2016-03-15     1.498525
2016-03-16     1.545111
2016-03-26     1.607226
2016-03-25     1.775454
2016-03-23     1.790983
2016-03-24     1.853098
2016-03-28     1.943682
2016-03-21     1.969564
2016-03-20     1.979916
2016-03-29     2.078265
2016-03-22     2.078265
2016-03-12     2.220612
2016-03-31     2.274962
2016-04-01     2.318961
2016-04-04     2.337077
2016-03-30     2.344842
2016-04-03     2.443191
2016-04-02     2.492365
2016-03-17     2.637300
2016-04-05    13.114033
2016-04-07    13.993996
2016-04-06    23.469124
Name: last_seen, dtype: float64

In [61]:
autos["registration_year"].describe()

count    38638.000000
mean      2005.676381
std         86.672400
min       1000.000000
25%       2001.000000
50%       2005.000000
75%       2009.000000
max       9999.000000
Name: registration_year, dtype: float64

While most of the data seems to fall within the right range, however there are a few outliers.
Max Year: 9999
Min Year: 1000

In [65]:
autos["registration_year"].value_counts().sort_index(ascending=True)

1000    1
1001    1
1927    1
1929    1
1931    1
       ..
5911    1
6200    1
8888    1
9000    1
9999    2
Name: registration_year, Length: 91, dtype: int64

Data shows that values out of the range are relatively less i.e is a good sign. Next, I would be removing the outliers from the dataset

In [68]:
autos = autos[autos["registration_year"].between(1900,2016)]
print(autos["registration_year"].value_counts(normalize=True)*100)

2005    7.483340
2006    7.125967
2004    7.007739
2003    6.655739
2007    6.069970
          ...   
1950    0.002687
1948    0.002687
1938    0.002687
1939    0.002687
1952    0.002687
Name: registration_year, Length: 77, dtype: float64


In [106]:
five_percent_and_more = autos["brand"].value_counts(normalize=True,ascending=False) > 0.005


counter = 0
top_brands = []
for s in five_percent_and_more:
    if s:
        top_brands.append(five_percent_and_more.index[counter])
    counter += 1

print(len(top_brands))


27


In [111]:
average_price_of_top_brands = {}

for brand in top_brands:
    avg_price = autos.loc[autos["brand"]==brand,"price"].mean()
    
    if brand not in average_price_of_top_brands:
        average_price_of_top_brands[brand] = round(avg_price,2)
        
for k,v in average_price_of_top_brands.items():
    print(k + ": " + str(v))

volkswagen: 8313.65
bmw: 9381.89
mercedes_benz: 9302.61
audi: 10322.27
opel: 7942.87
ford: 10866.91
renault: 3590.94
peugeot: 3955.17
fiat: 4008.17
skoda: 6836.7
seat: 5638.64
smart: 3780.47
toyota: 5573.57
mazda: 5309.53
citroen: 4614.97
nissan: 6428.43
mini: 10715.24
hyundai: 6181.87
sonstige_autos: 53093.93
volvo: 6151.58
kia: 6810.16
porsche: 46955.15
honda: 5297.28
mitsubishi: 4812.51
chevrolet: 7177.41
alfa_romeo: 5255.38
suzuki: 5149.88


In [117]:
average_mileage_of_top_cars = {}

for brand in top_brands:
    avg_mileage = autos.loc[autos["brand"]==brand,"odometer_km"].mean()
    
    if brand not in average_mileage_of_top_cars:
        average_mileage_of_top_cars[brand] = round(avg_mileage,2)


In [118]:
bma_series = pd.Series(average_mileage_of_top_cars)
bmp_series = pd.Series(average_price_of_top_brands)


In [119]:
df = pd.DataFrame(bma_series,columns=["mean_mileage"])
df["mean_price"] = bmp_series

In [120]:
df

mean_mileage  mean_price
volkswagen         125774.92     8313.65
bmw                132005.36     9381.89
mercedes_benz      130062.62     9302.61
audi               127491.05    10322.27
opel               123960.78     7942.87
ford               119590.58    10866.91
renault            121423.94     3590.94
peugeot            122341.04     3955.17
fiat               107901.79     4008.17
skoda              110063.47     6836.70
seat               116104.20     5638.64
smart               97775.08     3780.47
toyota             113740.81     5573.57
mazda              119981.13     5309.53
citroen            114245.65     4614.97
nissan             110335.31     6428.43
mini                88679.01    10715.24
hyundai            101862.50     6181.87
sonstige_autos      89438.78    53093.93
volvo              137230.54     6151.58
kia                107954.55     6810.16
porsche             98812.95    46955.15
honda              115989.01     5297.28
mitsubishi         119687.50     4812.51
chevrolet           98577.24     7177.41
alfa_romeo         129461.21     5255.38
suzuki             103755.87     5149.88